### ISSUE TITLE
Can we run Dataset API on GPU?

### ISSUE THREAD

**Issue by tongda** [tensorflow/tensorflow#13610]
I am running binary TF 1.3.0 on Ubuntu 16.04.

Python Version: 3.5.3

I have Nvidia TITAN Xp installed.

I use Dataset API to build input pipeline. What I want is to extract image features using CNN layers 
in the input pipeline. The code looks like this:

```
    def extract_feats(image):
      with tf.device("/gpu:0"):
        _, end_points = vgg.vgg_16(tf.expand_dims(image, 0),
                                   is_training=(mode == ModeKeys.TRAIN),
                                   spatial_squeeze=False)
        final_conv_layer = end_points['vgg_16/conv5/conv5_3']
        feats = spatial_pyramid_pooling(final_conv_layer, [bin_size], mode='avg')
      return tf.reshape(feats, shape=(bin_size * bin_size, tf.shape(final_conv_layer)[-1]))

    features = features.map(extract_feats)
```

When running the code, my CPU usage is more than 1000% (I have an 6 cores/12 threads CPU), while the GPU usage is 0%. I suspect that the input pipeline built from Dataset API are forced to run on CPU. I tried to set `log_device_placement=True` and I can see that the operation is placed on GPU.

Since I want to extract vectors with same length from variable-sized images using SPP pooling, I have to process these images one by one using `Dataset.map` before calling `Dataset.batch`. So I hope the operations inside `Dataset` could be run on GPU.

**Comment by asimshankar** [comment]
As you gathered, currently datasets execute all the computation on CPU.

@rohan100jain is in the process of adding some functionality to be able to invoke TensorFlow functions on different devices, which will be a pre-requisite but not sufficient for executing parts of the computation in a dataset pipeline on GPU.

So, long story short: At this moment there is no concrete timeline for executing the `map` function of a dataset on GPU.

FYI @mrry

**Comment by nirmalthacker** [comment]
Somewhat related question: I am curious how to effectively move data from the CPU to multiple GPU's using the Dataset API.
Do we need a dataset and iterator per GPU? Or can a single iterator provide unique data to each tower of the GPU? I have a more detailed question with an example code snippet here at stackoverflow: https://stackoverflow.com/questions/46965098/how-does-one-move-data-to-multiple-gpu-towers-using-tensorflows-dataset-api

**Comment by mrry** [comment]
@nirmalthacker  I posted an answer to your question [on Stack Overflow](https://stackoverflow.com/a/46966248/3574081).

@tongda General cross-device pipelines are still some way off, but @rohan100jain has developed some nice support for staging data automatically to GPU memory, which covers one of the big use cases. In principle you could reuse some of the support for dispatching TensorFlow functions to remote devices in a host-side `Dataset.map()`, but the partitioning between devices would currently be manual.

**Comment by nirmalthacker** [comment]
@mrry, Thanks! I've responded on Stackoverflow as well. In short, I'm trying to determine an approach with Datasets and Iterators that can provide the kind of throughput I'm seeing with the queue-based approach.

**Comment by staeff777** [comment]
I hope it's not too off-topic:
Is there a possibility to use efficient GPU-Preloading with the image_tensor placeholder in a object detection frozen graph? 

(I'm searching since two day, and think that datasets might be a good answer, however I'm afraid that the feed_dict parameter to set data for placeholders does not support any preloaded data.)

**Comment by agniszczotka** [comment]
I have noticed that GPU-utilization is lower when I use Dataset API. Is it possible to reach 80-90[%] utilization with this API or does native python threads perform worse then queues?

**Comment by mrry** [comment]
@agniszczotka It’s definitely possible to reach higher utilization, so it would be good to understand what the problem is here. One possibility is that the `Dataset` pipeline can prefetch and preprocess data more aggressively than the older API, which in some configurations can lead to oversubscription of the CPU, and thereby delay the dispatch of work to the GPU. Please open a new issue with a description of the program you’re running, and we’ll take a look.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by mrry** [comment]
No new status to report, but this is an area that we'll start to focus on in the new year.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Lancerchiang** [comment]
@mrry, Hi, I am recently building a multi-gpu model, does dataset API still do not support GPU well? I have been thinking if I should use dataset API or a queue-based approach, thanks!

**Comment by mrry** [comment]
@Lancerchaing Both the (current version of the) `tf.data` API and the old queue-based approach place the entire input pipeline on the CPU, and the GPUs are exercised by parts of the graph that come after the input pipeline. Since the overheads associated with the `tf.data` implementation are smaller, I'd recommend using it over the queue-based approach. Take a look at the [TF benchmarks](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py) for an example of how to use `tf.data` efficiently in a multi-GPU setup.

We still plan to add support for running stages of a `tf.data` pipeline on GPU, as well as easier-to-use support for prefetching to GPU memory (currently implemented [here](https://github.com/tensorflow/tensorflow/blob/926fc13f7378d14fa7980963c4fe774e5922e336/tensorflow/contrib/data/python/ops/prefetching_ops.py#L30), but not yet exposed via the API), which should help to achieve even better performance in future.

**Comment by Lancerchiang** [comment]
Hi thanks for the reply, I have looked into this benchmarking before and the other one here [https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10_estimator/cifar10_main.py](url) 
I am pretty new to tensorflow, and I need to change our existing project to a multi-gpu mode (instead of starting from scratch). Do you think below is a right approach with `tf.data` API ?

`
dataset = tf.data.TextLineDataset(filenames)
dataset = dataset.map(...) #preprocessing data
dataset = dataset.batch(batch_size)
iterator = dataset.make_initializable_iterator()

num_batches = len(filenames) // batch_size
predictions = []

for i in range(num_batches):
    batch = tf.split(dataset.next(), num_gpus) #splitting batch into smaller batches for each GPU
    for i in range(num_gpus):
        with tf.device('/gpu:%d' % i):
            with tf.name_scope('%s_%d' % (Tower_name, i)) as scope:
                predictions.append(inference(batch[i]))

`

Furthermore, do I need a runner like queue_runner in dataset API to facilitate the input pipeline?

**Comment by Lancerchiang** [comment]
@mrry Hi thanks for the reply, I have looked into this benchmarking before and the other one here [https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10_estimator/cifar10_main.py](url) 
I am pretty new to tensorflow, and I need to change our existing project to a multi-gpu mode (instead of starting from scratch). Do you think below is a right approach with `tf.data` API ?


```
dataset = tf.data.TextLineDataset(filenames)
dataset = dataset.map(...) #preprocessing data
dataset = dataset.batch(batch_size)
iterator = dataset.make_initializable_iterator()

num_batches = len(filenames) // batch_size
predictions = []

for i in range(num_batches):
    batch = tf.split(dataset.next(), num_gpus) #splitting batch into smaller batches for each GPU
    for i in range(num_gpus):
        with tf.device('/gpu:%d' % i):
            with tf.name_scope('%s_%d' % (Tower_name, i)) as scope:
                predictions.append(inference(batch[i]))

with tf.Session() as sess:
    sess.run(predictions)
```

Furthermore, do I need a runner like queue_runner in dataset API to facilitate the input pipeline （I hope I can to preprocess the next batch of image while GPU is doing computation）? 
Thanks :)

**Comment by ed-alertedh** [comment]
I'm looking forward to GPU prefetching as currently using StagingArea to manually pipeline data gets a little messy.

For the single-machine use-case I'm looking at using a feedable iterator to switch between training data and test data - is there any chance this could be supported on GPU so that it could be combined with prefetching? Trying to share the same prefetch pipeline between two data sources, while possible, would require very careful coordination to ensure the right data comes out at the right time. It would also be difficult to implement with existing high-level API's.

In crude diagram form, this is the approach I have in mind:
```
    CPU                                    | GPU
    train_data -> map -> prefetch<send> -> | -> prefetch<recv> -> string_handle\
                                                                                 -> feedable_iterator
    test_data  -> map -> prefetch<send> -> | -> prefetch<recv> -> string_handle/
```

**Comment by cipri-tom** [comment]
@tongda  how did you manage that problem in the end ? Are you really running the VGG on the CPU for preprocessing ?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by calledbymountains** [comment]
@TimZaman It cannot be used with initializable iterators which limits its usability. Isn't it ?

**Comment by sjperkins** [comment]
@calledbymountains I saw a commit to master that enables prefetch_to_device with initialisable iterators. I also think prefetch_to_device is in master, but not 1.7

**Comment by ychfan** [comment]
@mrry Is there any way to use `tf.contrib.data.prefetch_to_device` for mutl-gpu training? Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mrry** [comment]
Reassigning this to @rohan100jain, since he is currently looking into improvements to `prefetch_to_device()`, including for multi-GPU training and for chaining additional computation on the GPU.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rohan100jain** [comment]
Hi all,

Please take a look at MultiDeviceIterator https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/data/python/ops/prefetching_ops.py#L628

Currently its exposed via tf.contrib.data but this solves the prefetching to multiple GPU devices use case.

Sample API usage:

host_dataset_on_cpu = .....
multi_device_iterator = prefetching_ops.MultiDeviceIterator(host_dataset_on_cpu, devices=['/gpu:0', '/gpu:1'])
elem_on_gpu_0, elem_on_gpu_1 = multi_device_iterator.get_next()

with tf.Session() as sess:
  sess.run(multi_device_iterator.initializer)
  elem_0 = sess.run(elem_on_gpu_0)
  elem_1 = sess.run(elem_on_gpu_1)

MultiDeviceIterator deterministically prefetches elements to the different GPU's i.e. elements % 2 == 0 will go to GPU:0 and elements % 2 == 1 will go to GPU:1.

Closing this issue for now but feel free to open new ones if you see issues.

**Comment by jiazhe0909** [comment]
@rohan100jain Thanks for bringing this new iterator. Can I use it with Estimator?

**Comment by hollowgalaxy** [comment]
@rohan100jain does `to_device` support operations after.  This would be great to have since image preprocessing can be greatly accelerated on gpu.
